core/lib/vm_interface/src/pubdata/mod.rs

```diff
@@ -0,0 +1,90 @@
+use zksync_types::{
+    l2_to_l1_log::L2ToL1Log, writes::StateDiffRecord, Address, ProtocolVersionId, H256, U256,
+};
+
+/// Corresponds to the following solidity event:
+/// ```solidity
+/// struct L2ToL1Log {
+///     uint8 l2ShardId;
+///     bool isService;
+///     uint16 txNumberInBlock;
+///     address sender;
+///     bytes32 key;
+///     bytes32 value;
+/// }
+/// ```
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct L1MessengerL2ToL1Log {
+    pub l2_shard_id: u8,
+    pub is_service: bool,
+    pub tx_number_in_block: u16,
+    pub sender: Address,
+    pub key: U256,
+    pub value: U256,
+}
+
+impl L1MessengerL2ToL1Log {
+    pub fn packed_encoding(&self) -> Vec<u8> {
+        /// Converts `U256` value into bytes array
+        fn u256_to_bytes_be(value: &U256) -> Vec<u8> {
+            let mut bytes = vec![0u8; 32];
+            value.to_big_endian(bytes.as_mut_slice());
+            bytes
+        }
+
+        let mut res: Vec<u8> = vec![];
+        res.push(self.l2_shard_id);
+        res.push(self.is_service as u8);
+        res.extend_from_slice(&self.tx_number_in_block.to_be_bytes());
+        res.extend_from_slice(self.sender.as_bytes());
+        res.extend(u256_to_bytes_be(&self.key));
+        res.extend(u256_to_bytes_be(&self.value));
+        res
+    }
+}
+
+impl From<L1MessengerL2ToL1Log> for L2ToL1Log {
+    fn from(log: L1MessengerL2ToL1Log) -> Self {
+        fn u256_to_h256(num: U256) -> H256 {
+            let mut bytes = [0u8; 32];
+            num.to_big_endian(&mut bytes);
+            H256::from_slice(&bytes)
+        }
+
+        L2ToL1Log {
+            shard_id: log.l2_shard_id,
+            is_service: log.is_service,
+            tx_number_in_block: log.tx_number_in_block,
+            sender: log.sender,
+            key: u256_to_h256(log.key),
+            value: u256_to_h256(log.value),
+        }
+    }
+}
+
+/// Struct based on which the pubdata blob is formed
+#[derive(Debug, Clone, Default)]
+pub struct PubdataInput {
+    pub user_logs: Vec<L1MessengerL2ToL1Log>,
+    pub l2_to_l1_messages: Vec<Vec<u8>>,
+    pub published_bytecodes: Vec<Vec<u8>>,
```

Comment:
Dumb question / probably not important rn: Will this include EVM bytecodes if EVM emulator is enabled? (AFAIU, these bytecodes are at least known to the decommitment processor via [the EVM deployment tracer](https://github.com/matter-labs/zksync-era/blob/main/core/lib/multivm/src/versions/vm_latest/tracers/evm_deploy_tracer.rs).) *Should* it include such bytecodes?

Reply:
I've checked `stable/evm-emulator` branch in era-contracts, there `publishEVMBytecode` calls `L1_MESSENGER_CONTRACT.sendToL1` so these bytecodes will be included in `l2_to_l1_messages`. Historically we had a bit of a mess, compressed bytecodes were published to L1 through messages, uncompressed as `published_bytecodes` or vice versa I don't remember. It seems it has changed and now all eraVM bytecode go through `published_bytecodes`, I'm not sure why EVM bytecodes are different, cc @0xVolosnikov.
Overall it's not really important for node to which field bytecodes are included

Reply:
It would nice to switch to `published_bytecodes`, but i'm not sure what exactly is blocking us from this

@slowli any ideas?

Reply:
Changing `sendToL1` to `requestBytecodeL1Publication` [here](https://github.com/matter-labs/era-contracts/blob/stable/evm-emulator/system-contracts/contracts/KnownCodesStorage.sol#L101) should generally work

Reply:
I haven't had a chance to delve into it yet